### DISCUSSION TITLE
How to validate a JWT token server-side ?

### DISCUSSION THREAD

**Discussion by kevhof** [nuxt/nuxt#16454] [Questions]
I can't figure out how to validate a user JWT token before delivering them a component/page chunk ? 

I would very much appreciate your help:)

**Comment by Diizzayy** [comment]
@kay017 To start off, the server side needs to be aware of the JWT, hence it would need to be stored in a cookie that'll be passed to the server.

Providing that the aforementioned condition has been met, You'll be able to leverage a server side [Nuxt Plugin](https://v3.nuxtjs.org/guide/directory-structure/plugins) along with a [Nuxt Middleware](https://v3.nuxtjs.org/guide/directory-structure/middleware) to implement you desired authorization flow. Below is an example of what that might look like.

1. Server side auth plugin to validate JWT and initialize user state

```ts
import jwtDecode from 'jwt-decode'

`nuxt-app/plugins/auth.ts`
export default defineNuxtPlugin(() => {
  // handle user state, initialize as null
  const user = useState('user', () => null)

  const authCookie = useCookie('<jwt_cookie_name>')

  if (!authCookie.value) { return }

  // authCookie exists, validate it then extract data
  
  const decodedToken = jwtDecode(authCookie.value)

  // TODO: validate user based on the contents of your token ( `exp` perhaps ) or using an api request

  // update user state after validation
  user.value = {
    // set pertinent values from decodedToken
    id: decodedToken.id,
    role: decodedToken.role
  }
})
```

2. Global auth middleware to restrict access to desired pages

`nuxt-app/middleware/access.global.ts`
```ts
export default defineNuxtMiddleware((to, from) => {
  const user = useState('user')

  // require user with admin role to access '/admin'
  if (to.path === '/admin' && user.value?.role !== 'admin') {
    return navigateTo('/')
  }
})
```


Hopefully the outline above should be suffice to lead you in the right direction, feel free to ask any additional questions you may have

**Reply by danielrjames** [reply]
@Diizzayy it's a shame we can't use pinia in a plugin (in this instance), along with the entire pipeline breaking when you send an asynchronous request (renewing an expired jwt).

I solved this by moving this logic to app.vue (wrapped in a process.server if statement). The only cavet was to wrap the middleware with process.client (the pinia store isn't populated on initial server load).

Now is this the right way of doing things? Who knows, but it seems to work.

**Reply by Diizzayy** [reply]
@danielrjames I’m pretty sure pinia stores can indeed be used in plugins, once implemented properly it would circumvent your issue of the store not being populated on initial  load

**Reply by danielrjames** [reply]
@Diizzayy I'll have to check it out as I was getting errors back in rc-2. Perhaps there's been changes since then? I wonder if the nuxtapp instance being lost after async requests is still an issue as well?

**Reply by Diizzayy** [reply]
Oh yeah, the instance being lost after the first await still persists. That’s more so a vue 3 limitation as opposed to Nuxt

**Comment by kevhof** [comment]
@Diizzayy thank you for taking the time to answer. I'm sure this snippet you provided will be of great use later once I understand the concept. However I'm still confused at the moment. 

Aren't middlewares run only inside the user browser once the first page load is done ? If it's the case, the component is not really secure. A hacker could start fiddling with the code to go over the middleware, right?

**Reply by wobsoriano** [reply]
You are right, @kay017. Middlewares run only in the browser after the page load.
